text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+
+```rust
+let mut x: u32 = 3;
+unsafe {
+    asm!("add {0}, {number}", inout(reg) x, number = imm 5);
+}
+assert_eq!(x, 8);
+```
+
+We can see that `inout` is used to specify an argument that is both input and output.
+This is different from specifying an input and output separately in that it is guaranteed to assign both to the same register.
+
+It is also possible to specify different variables for the input and output parts of an `inout` operand:
+
+```rust
+let x: u32 = 3;
+let y: u32;
+unsafe {
+    asm!("add {0}, {number}", inout(reg) x => y, number = imm 5);
+}
+assert_eq!(y, 8);
+```
+
+## Late output operands
+
+The Rust compiler is conservative with its allocation of operands. It is assumed that an `out`
+can be written at any time, and can therefore not share its location with any other argument.
+However, to guarantee optimal performance it is important to use as few registers as possible,
+so they won't have to be saved and reloaded around the inline assembly block.
+To achieve this Rust provides a `lateout` specifier. This can be used on any output that is
+guaranteed to be written only after all inputs have been consumed.
+There is also a `inlateout` variant of this specifier.
+
+Here is an example where `inlateout` *cannot* be used:
+
+```rust
+let mut a = 4;
+let b = 4;
+let c = 4;
+unsafe {
+    asm!("
+        add {0}, {1}
+        add {0}, {2}
+    ", inout(reg) a, in(reg) b, in(reg) c);
+}
+assert_eq!(a, 12);
+```
+
+Here the compiler is free to allocate the same register for inputs `b` and `c` since it knows they have the same value. However it must allocate a separate register for `a` since it uses `inout` and not `inlateout`.
+
+However the following example can use `inlateout` since the output is only modified after all input registers have been read:
+
+```rust
+let mut a = 4;
+let b = 4;
+unsafe {
+    asm!("add {0}, {1}", inlateout(reg) a, in(reg) b);
+}
+assert_eq!(a, 8);
+```
+
+As you can see, this assembly fragment will still work correctly if `a` and `b` are assigned to the same register.
+
+## Explicit register operands
+
+Some instructions require that the operands be in a specific register.
+Therefore, Rust inline assembly provides some more specific constraint specifiers.
+While `reg` is generally available on any architecture, these are highly architecture specific. E.g. for x86 the general purpose registers `eax`, `ebx`, `ecx`, `edx`, `ebp`, `esi`, and `edi`
+among others can be addressed by their name.
+
+```rust
+unsafe {
+    asm!("out 0x64, {}", in("eax") cmd);
+}
+```
+
+In this example we call the `out` instruction to output the content of the `cmd` variable
+to port `0x64`. Since the `out` instruction only accepts `eax` (and its sub registers) as operand
+we had to use the `eax` constraint specifier.
+
+It is somewhat common that instructions have operands that are not explicitly listed in the
+assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:
```

Comment:
Implicit arguments are heavily used for syscalls, e.g.:
```rust
    asm!(
        "svc #0"
        : "={x0}" (result)
        : "{x8}" (nr as u64)
          "{x0}" (arg0)
          "{x1}" (arg1)
          "{x2}" (arg2)
          "{x3}" (arg3)
          "{x4}" (arg4)
          "{x5}" (arg5)
        : "x8", "memory"
        : "volatile"
    );
```

Something that was suggested during discussions was to lint against implicit arguments only if they don't use an explicit register.

Reply:
Actually I just realized that this is already stated in the reference section:

> The compiler will lint against any operands that are not used in the template string, except for operands that specify an explicit register.

Reply:
It sounds like the request is to also lint against using positional operands like `{}` if you're also using excess operands (like those with explicit registers). That doesn't seem like an unreasonable request, as a style lint that could be disabled (or perhaps enabled).

Reply:
What about requiring such operands to have a “name” of `_`?  e.g. `asm!(“svc #0”, _ = in(“x0”) arg0, /* etc. */)`

Reply:
@comex That syntax seems confusing to me; usually `_` means "ignored", not "unnamed".

How about the following, to replace the lint text @Amanieu referenced above:

"The compiler will lint against any operands that are not substituted in the template string, except for operands that specify an explicit register. If the format string contains explicit register operands not substituted in the template string, the compiler will lint against any operands substituted using positional arguments rather than named arguments."